src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.ConditionalExpressionInStringInterpolation
+{
+    internal partial class CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider
+    {
+        private sealed class AddParenthesisCodeAction : CodeAction
```

Comment:
General pattern is to not create a code action with logic in it.  But instead just subclass CodeAction.DocumentChangedCodeAction and just supply an appropriate lambda that does the actual work.  See other code actions for examples of this.